lib/sdl.cpp

```diff
@@ -0,0 +1,635 @@
+   if (GetMultisample() > 0)
+   {
+      SDL_GL_SetAttribute( SDL_GL_MULTISAMPLEBUFFERS, 1);
+      SDL_GL_SetAttribute( SDL_GL_MULTISAMPLESAMPLES, GetMultisample());
+      cerr << "Creating window..." << flush;
+   }
+   handle.reset(new Handle(title, x, y, w, h, win_flags));
+   if (GetMultisample() > 0 && !handle->isInitialized())
+   {
+      // Antialiasing support might not be available on all platforms.
+      PRINT_DEBUG("failed." << endl);
+      PRINT_DEBUG("Disabling antialiasing and trying again..." << flush);
+      SetMultisample(0);
+      SDL_GL_SetAttribute( SDL_GL_MULTISAMPLEBUFFERS, 0);
+      SDL_GL_SetAttribute( SDL_GL_MULTISAMPLESAMPLES, 0);
+      handle.reset(new Handle(title, x, y, w, h, win_flags));
+   }
+
+   // at this point, window should be up
+   if (!handle->isInitialized())
+   {
+      PRINT_DEBUG("failed." << endl);
+      cerr << "FATAL: window and/or OpenGL context creation failed." << endl;
+      return false;
+   }
+   else
+   {
+      PRINT_DEBUG("done." << endl);
+   }
+
+#ifndef __EMSCRIPTEN__
+   SDL_GL_SetSwapInterval(0);
+   glEnable(GL_DEBUG_OUTPUT);
+#endif
+
+   GLenum err = glewInit();
+   if (err != GLEW_OK)
+   {
+      cerr << "FATAL: Failed to initialize GLEW: "
+           << glewGetErrorString(err) << endl;
+      return false;
+   }
+
+   // print versions
+   PRINT_DEBUG("Using GLEW " << glewGetString(GLEW_VERSION) << std::endl);
+   PRINT_DEBUG("Using GL " << glGetString(GL_VERSION) << std::endl);
+
+   SDL_version sdl_ver;
+   SDL_GetVersion(&sdl_ver);
+   PRINT_DEBUG("Using SDL " << (int)sdl_ver.major << "." << (int)sdl_ver.minor
+               << "." << (int)sdl_ver.patch << std::endl);
+
+   renderer.reset(new gl3::MeshRenderer);
+#ifndef __EMSCRIPTEN__
+   if (GLEW_EXT_transform_feedback)
+   {
+      glBindBufferBase            = glBindBufferBaseEXT;
+      // Use an explicit typecast to suppress an error from inconsistent types
+      // that are present in older versions of GLEW.
+      glTransformFeedbackVaryings =
+         (PFNGLTRANSFORMFEEDBACKVARYINGSPROC)glTransformFeedbackVaryingsEXT;
+      glBeginTransformFeedback    = glBeginTransformFeedbackEXT;
+      glEndTransformFeedback      = glEndTransformFeedbackEXT;
+   }
+   if (!legacyGlOnly && (GLEW_VERSION_3_0
+                         || (GLEW_VERSION_2_0 && GLEW_EXT_transform_feedback)))
+   {
+      // We require both shaders and transform feedback EXT_transform_feedback
+      // was made core in OpenGL 3.0
+      PRINT_DEBUG("Loading CoreGLDevice..." << endl);
+      renderer->setDevice<gl3::CoreGLDevice>();
+   }
+   else
+   {
+      PRINT_DEBUG("Shader support missing, loading FFGLDevice..." << endl);
+      renderer->setDevice<gl3::FFGLDevice>();
+   }
+
+#else
+   renderer->setDevice<gl3::CoreGLDevice>();
+#endif
+
+   // Detect if we are using a high-dpi display and resize the window unless it
+   // was already resized by SDL's underlying backend.
+   {
```

Comment:
@v-dobrev: what's the main advantage of computing the DPI manually here instead of calling `SDL_GetDisplayDPI()`?

Reply:
The code is not computing dpi manually. It is calling `SDL_GetDisplayDPI()` through `getDpi()`, however, it first checks if SDL is already applying the window-size doubling for high-dpi displays, as it does e.g. under MacOS. If the "screen coordinate" and "real pixel" sizes of the window are different, we just assume the display is high-dpi and that SDL is already applying the doubling internally, so we do not need to do it (and there's no need to check the dpi here).

Reply:
Gotcha, that makes sense. Thanks!